pmda/parallel.py

```diff
@@ -307,144 +367,123 @@ def run(self,
-        self.timing = Timing(
-            np.hstack([el[1] for el in res]),
-            np.hstack([el[2] for el in res]), total.elapsed,
-            np.array([el[3] for el in res]), time_prepare,
-            conclude.elapsed,
-            # waiting time = wait_end - wait_start
-            np.array([el[4]-wait_start for el in res]),
-            np.array([el[5] for el in res]),
-            np.array([el[6] for el in res]))
         return self
 
-    def _dask_helper(self, bslice, indices, top, traj):
-        """helper function to actually setup dask graph"""
+    def _map_chunk(self, bslice):
+        """function to setup chunk dask graph"""
         # wait_end needs to be first line for accurate timing
         wait_end = time.time()
-        # record time to generate universe and atom groups
-        with timeit() as b_universe:
-            u = mda.Universe(top, traj)
-            agroups = [u.atoms[idx] for idx in indices]
-
         res = []
         times_io = []
         times_compute = []
         # NOTE: bslice.stop cannot be None! Always make sure
         #       that it comes from  _trajectory.check_slice_indices()!
         for i in range(bslice.start, bslice.stop, bslice.step):
+            self._frame_index = i
             # record io time per frame
             with timeit() as b_io:
                 # explicit instead of 'for ts in u.trajectory[bslice]'
                 # so that we can get accurate timing.
-                ts = u.trajectory[i]
+                self._ts = self._trajectory[i]
             # record compute time per frame
             with timeit() as b_compute:
-                res = self._reduce(res, self._single_frame(ts, agroups))
+                res = self._reduce(res, self._single_frame())
             times_io.append(b_io.elapsed)
             times_compute.append(b_compute.elapsed)
 
         # calculate io and compute time per block
         return np.asarray(res), np.asarray(times_io), np.asarray(
-            times_compute), b_universe.elapsed, wait_end, np.sum(
+            times_compute), wait_end, np.sum(
             times_io), np.sum(times_compute)
 
     @staticmethod
     def _reduce(res, result_single_frame):
         """ 'append' action for a time series"""
         res.append(result_single_frame)
         return res
+
+    def __getstate__(self):
+        #  To make sure Universe is pickled before Atomgroup,
+        base_dict = self.__dict__
+        universe_dict = {}
+        for key, item in base_dict.items():
+            if(isinstance(item, mda.Universe)):
+                universe_dict[key] = item
+        universe_dict.update(base_dict)
+        return universe_dict
+
+    #  dask-related functions
+
+    def __dask_graph__(self):
+        return self._dsk
+
+    def __dask_keys__(self):
+        return self._keys
+
+    #  it uses multiprocessing scheduler in default
+    __dask_scheduler__ = staticmethod(dask.multiprocessing.get)
```

Comment:
Even though multiprocessing is the default scheduler, one can still use distributed, right?

Reply:
Right, it can either be a global dask config, a context manager, or an arg in `self.compute()`. (https://docs.dask.org/en/latest/scheduler-overview.html#configuring-the-schedulers)